### ISSUE TITLE
Strapi does not properly use Unique field type with Bookshelf

### ISSUE THREAD

**Issue by derrickmehaffy** [strapi/strapi#1189]
**Informations**
- **Node.js version**: v9.11.1
- **npm version**: 6.0.0
- **Strapi version**: 3.0.0-alpha.12.1.3
- **Database**: 10.2.14-MariaDB
- **Operating system**: Ubuntu 16.04 LTS

**What is the current behavior?**

When setting a field as `unique` and using MySQL, strapi still allows for duplicate values to be written to the database. (Note that strapi does not create and apply a `UNIQUE` key on the specified column in the database)

**Steps to reproduce the problem**

Using MySQL or MariaDB:
Create a field in the content builder and assign the option `unique`, then attempt to save a value multiple times. Strapi will write multiple entries without throwing an error.

**What is the expected behavior?**

Strapi should (for MySQL, unsure about postgres) create and assign a UNIQUE key value to the specified field.

**Suggested solutions**

---

- [x] I'm sure that this feature hasn't already been referenced.

**Comment by derrickmehaffy** [comment]
postgres does allow for unique constraints which differs from MySQL:, but still allows for multiple `NULL` entires: https://www.postgresql.org/docs/9.3/static/indexes-unique.html

MySQL handles `unique` as an index and needs a key assigned (similar to the primary key) where `unique` values are indexed, allows for multiple `NULL` entires

MongoDB `unique` works, but not completely as intended as `NULL` values are considered unique when they shouldn't be.

**Comment by derrickmehaffy** [comment]
As a side note for relations it should in the future use foreign keys as well, while not a hard requirement, this should be done to keep in line with most database designs.

**Comment by derrickmehaffy** [comment]
https://dev.mysql.com/doc/refman/5.6/en/create-table-foreign-keys.html

**Comment by lauriejim** [comment]
Hello @, thank you for reporting this issue. Can you help us on this point, will appreciate your contribution!

**Comment by derrickmehaffy** [comment]
@lauriejim I'll see what I can do, I'm not really sure I completely understand how bookshelf itself works 😛

**Comment by derrickmehaffy** [comment]
@lauriejim it looks like this is something that needs to be modified in knex:

http://knexjs.org/#Schema-unique

http://knexjs.org/#Schema-foreign

**Comment by derrickmehaffy** [comment]
It looks like bookshelf is grabbing the `unique` variable from the model config to push to the database here:
https://github.com/strapi/strapi/blob/master/packages/strapi-bookshelf/lib/index.js#L400-L458

**Comment by derrickmehaffy** [comment]
After talking with @lauriejim this feature is known not to be implemented. In the mean time as a temp patch, you can manually modify your database to add unique key indexes as Strapi will not overwrite this.

The same should also apply to foreign keys as well as bookshelf/knex are not looking for, nor writing these keys.

**Comment by derrickmehaffy** [comment]
Modified title to better reflect my findings for all 3 databases: Mongo, Postgres, and MySQL

**Comment by lauriejim** [comment]
@derrickmehaffy what is the status of this issue please.

**Comment by derrickmehaffy** [comment]
I will double check, but as far as I'm aware keys are still not being made for the `Unique` option on MySQL

**Comment by derrickmehaffy** [comment]
Taking the follow field from one of my models:

```
    "systemName": {
      "type": "string",
      "required": true,
      "unique": true
    }
```

And poking at the database:
![image](https://user-images.githubusercontent.com/8593673/42140678-4cd17a8a-7d57-11e8-9c5e-52adaa7ba4d9.png)

You can see the only key defined is the `primary key` on the `id` column. Strapi should be generating `unique` keys on fields defined with the `unique` attribute.

Something like this: 
![image](https://user-images.githubusercontent.com/8593673/42140698-8484ab78-7d57-11e8-9819-c732f11a868d.png)

It should be noted that manually applying this to the database and attempting to POST duplicate values will cause an error in Strapi that is a 500 error. Meaning that bookshelf/knex are more or less ignoring the `unique` type attribute and this should be responding with something along the lines of "This value already exists" AKA possibly the 409 response code (typically used for PUT requests)

**Comment by derrickmehaffy** [comment]
Here is some documentation on the unique key constraint for MySQL: 

https://www.w3schools.com/sql/sql_unique.asp

**Comment by derrickmehaffy** [comment]
As for coding this into Bookshelf/Knex I'm afraid I don't really have the required skill set to develop this as I'm not terribly familiar with how bookshelf/knex work. :/

**Comment by derrickmehaffy** [comment]
Knex Unique function: https://knexjs.org/#Schema-unique

Bookshelf would need to grab the field meta `Unique Field` and run an `knex.schema.alterTable('$Tablename', function(t) {
  t.unique('$Columnname')`

**Comment by maturanomx** [comment]
I had seen this. I have in mind add it with Knexjs; lower --layer-- is better ;).

Specifically, can be added here: 
https://github.com/strapi/strapi/blob/637217a68fa21c3e36fe34f01609cb81dec1d6cb/packages/strapi-hook-bookshelf/lib/index.js#L575

**Comment by mamartins** [comment]
This is also not working with mongo

**Comment by amirakbulut** [comment]
How to solve this?

**Comment by derrickmehaffy** [comment]
@amirakbulut just add the constraint manually for now. Once the conversion of some of the custom queries to knex functions is complete then me and @maturanomx can work on keys and such

**Comment by amirakbulut** [comment]
> @amirakbulut just add the constraint manually for now. Once the conversion of some of the custom queries to knex functions is complete then me and @maturanomx can work on keys and such

So manually in let's say phpmyadmin?

**Comment by derrickmehaffy** [comment]
related: https://github.com/strapi/strapi/issues/2495

**Comment by kenberkeley** [comment]
Finally, I encountered this issue just now... I couldn't even believe this happened (violation of common sense 😂)

**Comment by panta82** [comment]
I "solved" this by adding the following to `beforeSave` hook on the model.

```
beforeSave: async (model, attrs, options) => {
  // Check that slug is unique
  if (attrs.slug) {
    const existing = await strapi.services.article.findOneBySlug(attrs.slug, false);
    if (existing && existing.get('id') !== model.get('id')) {
      // Not allowed
      throw new Error(
        `You can't set slug to "${
          attrs.slug
        }" because it is already taken by article #${existing.get('id')}`
      );
    }
  }
}
```

Unique key would be better, but I don't want to manually manage migrations of production database.

**Comment by fabiojansenbr** [comment]
Any fix or workaround to this bug? thanks

**Comment by vineshhpatel** [comment]
if somebody looking for handling error with custom message, Use following middleware in bootstrap.js file

```
module.exports = () => {
  strapi.app.use(async (ctx, next) => {
    try {
      await next();
    } catch(err) {
      if(err.code === 11000) {
        ctx.badData(`Duplicate value for ${Object.keys(err.keyPattern)}`);
      } else {
        throw err;
      }
    }
  });
};

```

I'm using MongoDB, Strapi 3.0.5

**Comment by aungkyawkyawsoe** [comment]
> if somebody looking for handling error with custom message, Use following middleware in bootstrap.js file
> 
> ```
> module.exports = () => {
>   strapi.app.use(async (ctx, next) => {
>     try {
>       await next();
>     } catch(err) {
>       if(err.code === 11000) {
>         ctx.badData(`Duplicate value for ${Object.keys(err.keyPattern)}`);
>       } else {
>         throw err;
>       }
>     }
>   });
> };
> ```
> 
> I'm using MongoDB, Strapi 3.0.5

@vineshhpatel  thanks for your valuable comments, by the way how could i check unique field for mongodb....

**Comment by itsmohmans** [comment]
Is this issue still not solved in v4? I'm encountering it in `4.12.0`
